Python/lock.c

```diff
@@ -353,3 +353,89 @@ _PyOnceFlag_CallOnceSlow(_PyOnceFlag *flag, _Py_once_fn_t *fn, void *arg)
         v = _Py_atomic_load_uint8(&flag->v);
     }
 }
+
+#define _PyRWMutex_READER_SHIFT 2
+
+void
+_PyRWMutex_RLock(_PyRWMutex *rwmutex)
+{
+    uintptr_t bits = _Py_atomic_load_uintptr_relaxed(&rwmutex->bits);
+    for (;;) {
+        // If the lock is not write-locked and there is no writer waiting, then
+        // we can increment the reader count.
+        if ((bits & (_Py_LOCKED|_Py_HAS_PARKED)) == 0) {
+            uintptr_t newval = bits + (1 << _PyRWMutex_READER_SHIFT);
```

Comment:
It may seem silly and unnecessary, but consider spelling out the constraint here.  For example:

```suggestion
        if ((bits & (_Py_LOCKED|_Py_HAS_PARKED)) == 0) {
            assert((bits >> _PyRWMutex_READER_SHIFT) < (UINTPTR_MAX >> _PyRWMutex_READER_SHIFT));
            uintptr_t newval = bits + (1 << _PyRWMutex_READER_SHIFT);
```

Reply:
I added an assertion.